### PR TITLE
Reduce docker image size by 18GB

### PR DESCRIPTION
**Pull request by basmaelgaabouri**
Reduce docker image size by 18.6GB
 
the changes consist of:

- splitting build depending on the infrastructure (SGX or TZ).
- changing the way rust and protobuf were installed inside the image and using a lighter way.
- using docker multistage build to reduce the build time.
- using docker buildkit to build different stages concurrently.
- using docker experiential feature squash to squash different image layers and make it possible to remove files from different layers.

Results of these changes are: 
for SGX: build time is 65s, size is 917MB.
for TZ: build time is 20min compared to 2h, size is 7.16GB compared to 25GB.
Screenshot 2020-11-24 at 20 14 27

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -11,29 +11,113 @@
...
+    && rm libsgx-enclave-common-dev_2.5.101.50123-bionic1_amd64.deb;\ 
+# Install PSW debug symbols
+    wget https://download.01.org/intel-sgx/linux-2.5/ubuntu18.04-server/libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && dpkg -i libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && rm libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb; \
+# more cleaning needed
+## fetch and build the SDK
+    wget https://download.01.org/intel-sgx/sgx-linux/2.9.1/distro/ubuntu18.04-server/sgx_linux_x64_sdk_2.9.101.2.bin \
+    && chmod a+x ./sgx_linux_x64_sdk_2.9.101.2.bin \
+    && echo "yes" | ./sgx_linux_x64_sdk_2.9.101.2.bin \
```
Comment (hugovincent) [inline Dockerfile:53]: Not your fault but do they really distribute the SDK as a self-extracting executable? Yuck!

Comment (basmaelgaabouri) [reply Dockerfile:53]: in their [documentation](https://download.01.org/intel-sgx/sgx-linux/2.9.1/docs/Intel_SGX_Installation_Guide_Linux_2.9.1_Open_Source.pdf) that's still the official recommendation, but we can change that to building the executable from their git repo, should I change that?

Comment (hugovincent) [reply Dockerfile:53]: I wasn't requesting a change, just making a comment – if that's what they recommend it's fine (just a little surprising to see in a security-oriented project).

Comment (ShaleXIONG) [reply Dockerfile:53]: They also publish source code [here](https://github.com/intel/linux-sgx) @hugovincent.

**Comment by dominic-mulligan-arm** [comment]
Great work, @basmaelgaabouri.  I've just merged @geky's pull request ahead of this one, as it was apparently sat in the queue for two weeks without anybody noticing.  From looking at his changes, it shouldn't be too hard to rebase your changes on top of his: he just adds `absname` to the Makefile in various spots around `VERACRUZ_ROOT` so that relative paths work.  This looks good to me, though I'll wait for a review from @ShaleXIONG before merging.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# Veracruz Docker image
+
+## Requierments
```
Comment (ebeasant-arm) [inline README.md:3]: Very nitty nit, typo: ## Requirements

File: Makefile [hunk 3]
```diff
@@ -22,26 +22,27 @@ development: build
...
 .PHONY:
 build: Dockerfile
-	docker build --build-arg USER=$(USER) --build-arg UID=$(UID) -t $(VERACRUZ_DOCKER_IMAGE) -f $< .
+	DOCKER_BUILDKIT=1 docker build --squash --build-arg USER=$(USER) --build-arg UID=$(UID) --build-arg TEE=$(TEE) -t $(VERACRUZ_DOCKER_IMAGE)_$(TEE) -f $< .
 
 .PHONY:
 ci: Dockerfile.ci
-	docker build --build-arg USER=root --build-arg UID=0 -t $(VERACRUZ_DOCKER_IMAGE) -f $<  .
+	DOCKER_BUILDKIT=1 docker build --squash --build-arg USER=root --build-arg UID=0 --build-arg TEE=sgx -t $(VERACRUZ_DOCKER_IMAGE)_sgx -f $<  .
+	DOCKER_BUILDKIT=1 docker build --squash --build-arg USER=root --build-arg UID=0 --build-arg TEE=tz -t $(VERACRUZ_DOCKER_IMAGE)_tz -f $<  .
```
Comment (ebeasant-arm) [inline Makefile:45]: Where a multi-layer build is used (in this case for trustzone), I might have expected to see a --cache-from=build_tz_base and perhaps a --cache-from=build_tee here?

Comment (basmaelgaabouri) [reply Makefile:45]: to my understanding, it's done implicitly since some stages depend on the one before, I could be wrong though ..

Comment (ebeasant-arm) [reply Makefile:50]: If it shows up as a problem after the commit, we can revisit it, I guess :-)

File: Dockerfile [hunk 4]
```diff
@@ -11,29 +11,113 @@
...
+# Install PSW-dev
+WORKDIR /work
+RUN wget https://download.01.org/intel-sgx/linux-2.5/ubuntu18.04-server/libsgx-enclave-common-dev_2.5.101.50123-bionic1_amd64.deb \
+    && dpkg -i libsgx-enclave-common-dev_2.5.101.50123-bionic1_amd64.deb \
+    && rm libsgx-enclave-common-dev_2.5.101.50123-bionic1_amd64.deb;\ 
+# Install PSW debug symbols
+    wget https://download.01.org/intel-sgx/linux-2.5/ubuntu18.04-server/libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && dpkg -i libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb \
+    && rm libsgx-enclave-common-dbgsym_2.5.101.50123-bionic1_amd64.ddeb; \
+# more cleaning needed
```
Comment (dominic-mulligan-arm) [inline Dockerfile:49]: Is this comment stray?  Or is it a TODO?

File: README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
+# Veracruz Docker image
+
+## Requirements
+
+We use `squash` docker experimental feature to help reduce docker image size, to enable this feature:
+- run:
```
Comment (dominic-mulligan-arm) [inline README.md:6]: Are these instructions Linux-specific?  We should probably note that, if they are.

Comment (basmaelgaabouri) [reply README.md:8]: yes they are, I've updated readme

File: Makefile (hunk 3 above)
Comment (basmaelgaabouri) [reply Makefile:50]: sounds good to me

File: Dockerfile (hunk 4 above)
Comment (basmaelgaabouri) [reply Dockerfile:49]: oops! .. forgot it there

**Review by dominic-mulligan-arm — APPROVED** [review]

File: Dockerfile [hunk 6]
```diff
@@ -115,97 +198,84 @@ RUN apt-get update && apt-get install -y \
...
+RUN chown -R $USER /work
+USER $USER
 RUN git clone --depth 1 --branch veracruz https://github.com/veracruz-project/rust-optee-trustzone-sdk.git \
     && cd rust-optee-trustzone-sdk \ 
     && git submodule update --init \
-    && cd rust/compiler-builtins \
-    && git submodule update --init libm \
-    && cd ../rust \
-    && git submodule update --init src/stdarch src/llvm-project 
+    && rm -rf rust
```
[the remark this answers is not in the document]
Comment (basmaelgaabouri) [reply Dockerfile:259]: fixed

File: Dockerfile [hunk 7]
```diff
@@ -11,29 +11,113 @@
...
 ENV DEBIAN_FRONTEND noninteractive
+# Use bash as the default
+SHELL ["/bin/bash", "-c"]
 
-# In root, install dependencies and set up local user
-USER root
+RUN apt-get update && apt-get install --no-install-recommends -y \
+    libcurl4-openssl-dev libprotobuf-dev python \
+    libssl-dev build-essential \
+    wget make curl unzip ca-certificates \
```
[the remark this answers is not in the document]
Comment (basmaelgaabouri) [reply Dockerfile:27]: fixed

File: Dockerfile [hunk 8]
```diff
@@ -115,97 +198,84 @@ RUN apt-get update && apt-get install -y \
...
+    chmod +x rustup-init; \
+    ./rustup-init -y --no-modify-path --profile minimal --default-toolchain $RUST_VERSION --default-host ${rustArch}; \
+    rm rustup-init; \
+    chmod -R a+w $RUSTUP_HOME $CARGO_HOME; \
+    # Install the dependencies for rust,
 # seting up for cross-compiling for aarch64 and trustzone
-RUN if [ "$USER" != "root" ] ; then . /home/$USER/.cargo/env ; else . /root/.cargo/env; fi \
+    if [ "$USER" != "root" ] ; then . /home/$USER/.cargo/env ; else . /root/.cargo/env; fi \
     && rustup component add rust-src \
     && cargo install xargo --force \
```
[the remark this answers is not in the document]
Comment (basmaelgaabouri) [reply Dockerfile:247]: fixed

**Review by ShaleXIONG — APPROVED** [review]

**Comment by dominic-mulligan-arm** [comment]
@ShaleXIONG to confirm, the changes in the main Veracruz repository to support this change can be merged both before or after this is approved, right?  That is, there is no dependency on either one being merged first?

If so, @dreemkiller, when you are happy with these changes please go ahead and merge these changes.

**Comment by ShaleXIONG** [comment]
> @ShaleXIONG to confirm, the changes in the main Veracruz repository to support this change can be merged both before or after this is approved, right? That is, there is no dependency on either one being merged first?
> 
> If so, @dreemkiller, when you are happy with these changes please go ahead and merge these changes.

@dominic-mulligan-arm . There is no direct dependency. I will wait this go through first, and then I will replace the image in AWS. Therefore we can let CI check if [the veracruz pull request](https://github.com/veracruz-project/veracruz/pull/52) is good. However, without the new change, the old Veracruz repo will NOT be able to compile in the new docker image. That is, the new image has a 'smaller' environment.

**Comment by dreemkiller** [comment]
When I build for SGX, I get `execvp: clang-check: Permission denied` when building psa-attestation. Is anyone else seeing this?

**Comment by ShaleXIONG** [comment]
> When I build for SGX, I get `execvp: clang-check: Permission denied` when building psa-attestation. Is anyone else seeing this?

Short answer: I did not see this before.
I used `make sgx` in my run. Did you manually compile psa-attestaion and get that problem? Can you provide a bit more information, so @basmaelgaabouri and I can check what goes wrong in the image.

**Comment by dreemkiller** [comment]
I used `make sgx-sinaloa-test`

**Comment by dreemkiller** [comment]
Ok, I think I have more germane questions.
When I enter the docker container with my username (`docker exec -u <USERNAME> -it veracruz bash`), and I run `make sgx-sinaloa-test`, I get the following error:

> make -C data-generators/string-generator/
make[2]: Entering directory '/work/veracruz/sdk/data-generators/string-generator'
cargo run -- --input_file hello-world-1.txt
info: syncing channel updates for 'nightly-2020-05-07-x86_64-unknown-linux-gnu'
info: latest update on 2020-05-07, rust version 1.45.0-nightly (1836e3b42 2020-05-06)
info: downloading component 'cargo'
info: downloading component 'rust-std'
info: downloading component 'rustc'
info: installing component 'cargo'
info: Defaulting to 500.0 MiB unpack ram
info: installing component 'rust-std'
info: installing component 'rustc'
    Updating crates.io index
warning: spurious network error (2 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_9tRRf3': Permission denied; class=Os (2)
warning: spurious network error (1 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_UtETaF': Permission denied; class=Os (2)
error: failed to get `clap` as a dependency of package `string-generator v0.3.0 (/work/veracruz/sdk/data-generators/string-generator)`
Caused by:
  failed to load source for dependency `clap`
Caused by:
  Unable to update registry `https://github.com/rust-lang/crates.io-index`
Caused by:
  failed to fetch `https://github.com/rust-lang/crates.io-index`
Caused by:
  failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_MtgSpi': Permission denied; class=Os (2)
Makefile:29: recipe for target 'hello-world-1.dat' failed

This appears to be a problem with which user is active when rust was installed in the Dockerfile. Previously, the` $USER `was active. In the new version, the active user is `root`. Thus, previously cargo and rust were installed in /home/$USER/.cargo, but are now installed in `/usr/local/cargo`, which is not writable by the `$USER` user.
I'm not sure how to resolve this without possibly breaking the CI system, so I need some advice on why this change was made and how we might resolve it.

**Comment by ShaleXIONG** [comment]
> Ok, I think I have more germane questions.
> When I enter the docker container with my username (`docker exec -u <USERNAME> -it veracruz bash`), and I run `make sgx-sinaloa-test`, I get the following error:
> 
> > make -C data-generators/string-generator/
> > make[2]: Entering directory '/work/veracruz/sdk/data-generators/string-generator'
> > cargo run -- --input_file hello-world-1.txt
> > info: syncing channel updates for 'nightly-2020-05-07-x86_64-unknown-linux-gnu'
> > info: latest update on 2020-05-07, rust version 1.45.0-nightly (1836e3b42 2020-05-06)
> > info: downloading component 'cargo'
> > info: downloading component 'rust-std'
> > info: downloading component 'rustc'
> > info: installing component 'cargo'
> > info: Defaulting to 500.0 MiB unpack ram
> > info: installing component 'rust-std'
> > info: installing component 'rustc'
> > Updating crates.io index
> > warning: spurious network error (2 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_9tRRf3': Permission denied; class=Os (2)
> > warning: spurious network error (1 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_UtETaF': Permission denied; class=Os (2)
> > error: failed to get `clap` as a dependency of package `string-generator v0.3.0 (/work/veracruz/sdk/data-generators/string-generator)`
> > Caused by:
> > failed to load source for dependency `clap`
> > Caused by:
> > Unable to update registry `https://github.com/rust-lang/crates.io-index`
> > Caused by:
> > failed to fetch `https://github.com/rust-lang/crates.io-index`
> > Caused by:
> > failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_MtgSpi': Permission denied; class=Os (2)
> > Makefile:29: recipe for target 'hello-world-1.dat' failed
> 
> This appears to be a problem with which user is active when rust was installed in the Dockerfile. Previously, the`$USER`was active. In the new version, the active user is `root`. Thus, previously cargo and rust were installed in /home/$USER/.cargo, but are now installed in `/usr/local/cargo`, which is not writable by the `$USER` user.
> I'm not sure how to resolve this without possibly breaking the CI system, so I need some advice on why this change was made and how we might resolve it.

One solution is to grant permission to $USER for accessing /usr/local/cargo, when building up the image. @basmaelgaabouri Can you make the change?

**Comment by dominic-mulligan-arm** [comment]
@basmaelgaabouri, @ShaleXIONG, @dreemkiller, @ebeasant-arm: this now looks ready for merge.  Any final comments/blockers on this?

**Review by ShaleXIONG — APPROVED** [review]

**Pull request merged**